File: backend/app/routers/resources.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.models import (
    Resource, ResourceAllocation, ResourceEvent,
    ResourceETA, ResourceAssignment, Supplier
)
from app.schemas.resources.resource import ResourceCreate, ResourceUpdate, ResourceOut, ResourceWithDetails
from app.schemas.resources.allocation import ResourceAllocationCreate, ResourceAllocationUpdate, ResourceAllocationOut
from app.schemas.resources.event import ResourceEventCreate, ResourceEventUpdate, ResourceEventOut
from app.schemas.resources.eta import ResourceETACreate, ResourceETAUpdate, ResourceETAOut
from app.schemas.resources.assignment import ResourceAssignmentCreate, ResourceAssignmentUpdate, ResourceAssignmentOut
from app.schemas.resources.supplier import SupplierCreate, SupplierUpdate, SupplierOut

router = APIRouter(prefix="/resources", tags=["Resources"])


# ── Suppliers ─────────────────────────────────────────────

@router.get("/suppliers", response_model=list[SupplierOut])
def get_all_suppliers(db: Session = Depends(get_db)):
    return db.query(Supplier).all()


@router.get("/suppliers/{supplier_id}", response_model=SupplierOut)
def get_supplier(supplier_id: int, db: Session = Depends(get_db)):
    obj = db.query(Supplier).filter(Supplier.id == supplier_id).first()
    if not obj:
        raise HTTPException(status_code=404, detail="Supplier not found")
    return obj


@router.post("/suppliers", response_model=SupplierOut, status_code=201)
def create_supplier(data: SupplierCreate, db: Session = Depends(get_db)):
    obj = Supplier(**data.model_dump())
    db.add(obj)
    db.commit()
    db.refresh(obj)
    return obj


@router.patch("/suppliers/{supplier_id}", response_model=SupplierOut)
def update_supplier(supplier_id: int, data: SupplierUpdate, db: Session = Depends(get_db)):
    obj = db.query(Supplier).filter(Supplier.id == supplier_id).first()
    if not obj:
        raise HTTPException(status_code=404, detail="Supplier not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(obj, field, value)
    db.commit()
    db.refresh(obj)
    return obj


@router.delete("/suppliers/{supplier_id}", status_code=204)
def delete_supplier(supplier_id: int, db: Session = Depends(get_db)):
    obj = db.query(Supplier).filter(Supplier.id == supplier_id).first()
    if not obj:
        raise HTTPException(status_code=404, detail="Supplier not found")
    db.delete(obj)
    db.commit()


# ── Resources ─────────────────────────────────────────────

@router.get("/", response_model=list[ResourceOut])
def get_all_resources(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    return db.query(Resource).offset(skip).limit(limit).all()


@router.get("/{resource_id}", response_model=ResourceWithDetails)
def get_resource(resource_id: int, db: Session = Depends(get_db)):
    obj = db.query(Resource).filter(Resource.id == resource_id).first()
    if not obj:
        raise HTTPException(status_code=404, detail="Resource not found")
    return obj


@router.post("/", response_model=ResourceOut, status_code=201)
def create_resource(data: ResourceCreate, db: Session = Depends(get_db)):
    obj = Resource(**data.model_dump())
    db.add(obj)
    db.commit()
    db.refresh(obj)
    return obj


@router.patch("/{resource_id}", response_model=ResourceOut)
def update_resource(resource_id: int, data: ResourceUpdate, db: Session = Depends(get_db)):
    obj = db.query(Resource).filter(Resource.id == resource_id).first()
    if not obj:
        raise HTTPException(status_code=404, detail="Resource not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(obj, field, value)
    db.commit()
    db.refresh(obj)
    return obj


@router.delete("/{resource_id}", status_code=204)
def delete_resource(resource_id: int, db: Session = Depends(get_db)):
    obj = db.query(Resource).filter(Resource.id == resource_id).first()
    if not obj:
        raise HTTPException(status_code=404, detail="Resource not found")
    db.delete(obj)
    db.commit()


# ── Resource Allocations ──────────────────────────────────

@router.get("/{resource_id}/allocations", response_model=list[ResourceAllocationOut])
def get_allocations(resource_id: int, db: Session = Depends(get_db)):
    return db.query(ResourceAllocation).filter(ResourceAllocation.resource_id == resource_id).all()


@router.post("/allocations", response_model=ResourceAllocationOut, status_code=201)
def create_allocation(data: ResourceAllocationCreate, db: Session = Depends(get_db)):
    obj = ResourceAllocation(**data.model_dump())
    db.add(obj)
    db.commit()
    db.refresh(obj)
    return obj


@router.patch("/allocations/{allocation_id}", response_model=ResourceAllocationOut)
def update_allocation(allocation_id: int, data: ResourceAllocationUpdate, db: Session = Depends(get_db)):
    obj = db.query(ResourceAllocation).filter(ResourceAllocation.id == allocation_id).first()
    if not obj:
        raise HTTPException(status_code=404, detail="Allocation not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(obj, field, value)
    db.commit()
    db.refresh(obj)
    return obj


@router.delete("/allocations/{allocation_id}", status_code=204)
def delete_allocation(allocation_id: int, db: Session = Depends(get_db)):
    obj = db.query(ResourceAllocation).filter(ResourceAllocation.id == allocation_id).first()
    if not obj:
        raise HTTPException(status_code=404, detail="Allocation not found")
    db.delete(obj)
    db.commit()


# ── Resource Events ───────────────────────────────────────

@router.get("/{resource_id}/events", response_model=list[ResourceEventOut])
def get_events(resource_id: int, db: Session = Depends(get_db)):
    return db.query(ResourceEvent).filter(ResourceEvent.resource_id == resource_id).all()


@router.post("/events", response_model=ResourceEventOut, status_code=201)
def create_event(data: ResourceEventCreate, db: Session = Depends(get_db)):
    obj = ResourceEvent(**data.model_dump())
    db.add(obj)
    db.commit()
    db.refresh(obj)
    return obj


@router.patch("/events/{event_id}", response_model=ResourceEventOut)
def update_event(event_id: int, data: ResourceEventUpdate, db: Session = Depends(get_db)):
    obj = db.query(ResourceEvent).filter(ResourceEvent.id == event_id).first()
    if not obj:
        raise HTTPException(status_code=404, detail="Event not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(obj, field, value)
    db.commit()
    db.refresh(obj)
    return obj


@router.delete("/events/{event_id}", status_code=204)
def delete_event(event_id: int, db: Session = Depends(get_db)):
    obj = db.query(ResourceEvent).filter(ResourceEvent.id == event_id).first()
    if not obj:
        raise HTTPException(status_code=404, detail="Event not found")
    db.delete(obj)
    db.commit()


# ── Resource ETA ──────────────────────────────────────────

@router.get("/{resource_id}/eta", response_model=list[ResourceETAOut])
def get_eta(resource_id: int, db: Session = Depends(get_db)):
    return db.query(ResourceETA).filter(ResourceETA.resource_id == resource_id).all()


@router.post("/eta", response_model=ResourceETAOut, status_code=201)
def create_eta(data: ResourceETACreate, db: Session = Depends(get_db)):
    obj = ResourceETA(**data.model_dump())
    db.add(obj)
    db.commit()
    db.refresh(obj)
    return obj


@router.patch("/eta/{eta_id}", response_model=ResourceETAOut)
def update_eta(eta_id: int, data: ResourceETAUpdate, db: Session = Depends(get_db)):
    obj = db.query(ResourceETA).filter(ResourceETA.id == eta_id).first()
    if not obj:
        raise HTTPException(status_code=404, detail="ETA record not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(obj, field, value)
    db.commit()
    db.refresh(obj)
    return obj


@router.delete("/eta/{eta_id}", status_code=204)
def delete_eta(eta_id: int, db: Session = Depends(get_db)):
    obj = db.query(ResourceETA).filter(ResourceETA.id == eta_id).first()
    if not obj:
        raise HTTPException(status_code=404, detail="ETA record not found")
    db.delete(obj)
    db.commit()


# ── Resource Assignments ──────────────────────────────────

@router.get("/{resource_id}/assignments", response_model=list[ResourceAssignmentOut])
def get_assignments(resource_id: int, db: Session = Depends(get_db)):
    return db.query(ResourceAssignment).filter(ResourceAssignment.resource_id == resource_id).all()


@router.post("/assignments", response_model=ResourceAssignmentOut, status_code=201)
def create_assignment(data: ResourceAssignmentCreate, db: Session = Depends(get_db)):
    obj = ResourceAssignment(**data.model_dump())
    db.add(obj)
    db.commit()
    db.refresh(obj)
    return obj


@router.patch("/assignments/{assignment_id}", response_model=ResourceAssignmentOut)
def update_assignment(assignment_id: int, data: ResourceAssignmentUpdate, db: Session = Depends(get_db)):
    obj = db.query(ResourceAssignment).filter(ResourceAssignment.id == assignment_id).first()
    if not obj:
        raise HTTPException(status_code=404, detail="Assignment not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(obj, field, value)
    db.commit()
    db.refresh(obj)
    return obj


@router.delete("/assignments/{assignment_id}", status_code=204)
def delete_assignment(assignment_id: int, db: Session = Depends(get_db)):
    obj = db.query(ResourceAssignment).filter(ResourceAssignment.id == assignment_id).first()
    if not obj:
        raise HTTPException(status_code=404, detail="Assignment not found")
    db.delete(obj)
    db.commit()

File: frontend/lib/api/resources.ts
/**
 * lib/api/resources.ts
 * Covers: suppliers, resources, allocations, events, ETAs, assignments
 */

import { apiRequest } from "./client";

// ── Supplier ──────────────────────────────────────────────

export interface SupplierOut {
  id: number;
  name: string;
  initials?: string;
  contact_info?: string;
}

export interface SupplierCreate {
  name: string;
  initials?: string;
  contact_info?: string;
}

export interface SupplierUpdate {
  name?: string;
  initials?: string;
  contact_info?: string;
}

export async function getSuppliers(): Promise<SupplierOut[]> {
  return apiRequest<SupplierOut[]>("/resources/suppliers");
}

export async function getSupplier(supplierId: number): Promise<SupplierOut> {
  return apiRequest<SupplierOut>(`/resources/suppliers/${supplierId}`);
}

export async function createSupplier(data: SupplierCreate): Promise<SupplierOut> {
  return apiRequest<SupplierOut>("/resources/suppliers", { method: "POST", body: data });
}

export async function updateSupplier(supplierId: number, data: SupplierUpdate): Promise<SupplierOut> {
  return apiRequest<SupplierOut>(`/resources/suppliers/${supplierId}`, { method: "PATCH", body: data });
}

export async function deleteSupplier(supplierId: number): Promise<void> {
  return apiRequest<void>(`/resources/suppliers/${supplierId}`, { method: "DELETE" });
}

// ── Resource ──────────────────────────────────────────────

export interface ResourceOut {
  id: number;
  name: string;
  type: string;
  code?: string;
  supplier_id?: number;
  quantity?: number;
  unit?: string;
  status?: string;
  created_at?: string;
}

export interface ResourceWithDetails extends ResourceOut {
  supplier?: SupplierOut | null;
}

export interface ResourceCreate {
  name: string;
  type: string;
  code?: string;
  supplier_id?: number;
  quantity?: number;
  unit?: string;
  status?: string;
}

export interface ResourceUpdate {
  name?: string;
  type?: string;
  code?: string;
  supplier_id?: number;
  quantity?: number;
  unit?: string;
  status?: string;
}

export async function getResources(skip = 0, limit = 100): Promise<ResourceOut[]> {
  return apiRequest<ResourceOut[]>("/resources", { params: { skip, limit } });
}

export async function getResource(resourceId: number): Promise<ResourceWithDetails> {
  return apiRequest<ResourceWithDetails>(`/resources/${resourceId}`);
}

export async function createResource(data: ResourceCreate): Promise<ResourceOut> {
  return apiRequest<ResourceOut>("/resources", { method: "POST", body: data });
}

export async function updateResource(resourceId: number, data: ResourceUpdate): Promise<ResourceOut> {
  return apiRequest<ResourceOut>(`/resources/${resourceId}`, { method: "PATCH", body: data });
}

export async function deleteResource(resourceId: number): Promise<void> {
  return apiRequest<void>(`/resources/${resourceId}`, { method: "DELETE" });
}

// ── Allocation ────────────────────────────────────────────

export interface AllocationOut {
  id: number;
  resource_id: number;
  project_id: number;
  quantity_used?: number;
  phase?: string;
  status?: string;
  assigned_at?: string;
}

export interface AllocationCreate {
  resource_id: number;
  project_id: number;
  quantity_used?: number;
  phase?: string;
  status?: string;
}

export interface AllocationUpdate {
  quantity_used?: number;
  phase?: string;
  status?: string;
}

export async function getAllocations(resourceId: number): Promise<AllocationOut[]> {
  return apiRequest<AllocationOut[]>(`/resources/${resourceId}/allocations`);
}

export async function createAllocation(data: AllocationCreate): Promise<AllocationOut> {
  return apiRequest<AllocationOut>("/resources/allocations", { method: "POST", body: data });
}

export async function updateAllocation(allocationId: number, data: AllocationUpdate): Promise<AllocationOut> {
  return apiRequest<AllocationOut>(`/resources/allocations/${allocationId}`, { method: "PATCH", body: data });
}

export async function deleteAllocation(allocationId: number): Promise<void> {
  return apiRequest<void>(`/resources/allocations/${allocationId}`, { method: "DELETE" });
}

// ── Event ─────────────────────────────────────────────────

export interface ResourceEventOut {
  id: number;
  resource_id: number;
  event_type?: string;
  progress?: number;
  note?: string;
  event_date?: string;
}

export interface ResourceEventCreate {
  resource_id: number;
  event_type?: string;
  progress?: number;
  note?: string;
}

export interface ResourceEventUpdate {
  event_type?: string;
  progress?: number;
  note?: string;
}

export async function getResourceEvents(resourceId: number): Promise<ResourceEventOut[]> {
  return apiRequest<ResourceEventOut[]>(`/resources/${resourceId}/events`);
}

export async function createResourceEvent(data: ResourceEventCreate): Promise<ResourceEventOut> {
  return apiRequest<ResourceEventOut>("/resources/events", { method: "POST", body: data });
}

export async function updateResourceEvent(eventId: number, data: ResourceEventUpdate): Promise<ResourceEventOut> {
  return apiRequest<ResourceEventOut>(`/resources/events/${eventId}`, { method: "PATCH", body: data });
}

export async function deleteResourceEvent(eventId: number): Promise<void> {
  return apiRequest<void>(`/resources/events/${eventId}`, { method: "DELETE" });
}

// ── ETA ───────────────────────────────────────────────────

export interface ResourceETAOut {
  id: number;
  resource_id: number;
  eta_date?: string;
  status_text?: string;
  sub_text?: string;
}

export interface ResourceETACreate {
  resource_id: number;
  eta_date?: string;
  status_text?: string;
  sub_text?: string;
}

export interface ResourceETAUpdate {
  eta_date?: string;
  status_text?: string;
  sub_text?: string;
}

export async function getResourceETAs(resourceId: number): Promise<ResourceETAOut[]> {
  return apiRequest<ResourceETAOut[]>(`/resources/${resourceId}/eta`);
}

export async function createResourceETA(data: ResourceETACreate): Promise<ResourceETAOut> {
  return apiRequest<ResourceETAOut>("/resources/eta", { method: "POST", body: data });
}

export async function updateResourceETA(etaId: number, data: ResourceETAUpdate): Promise<ResourceETAOut> {
  return apiRequest<ResourceETAOut>(`/resources/eta/${etaId}`, { method: "PATCH", body: data });
}

export async function deleteResourceETA(etaId: number): Promise<void> {
  return apiRequest<void>(`/resources/eta/${etaId}`, { method: "DELETE" });
}

// ── Assignment ────────────────────────────────────────────

export interface ResourceAssignmentOut {
  id: number;
  resource_id: number;
  user_id: number;
  role?: string;
  start_time?: string;
  end_time?: string;
}

export interface ResourceAssignmentCreate {
  resource_id: number;
  user_id: number;
  role?: string;
  start_time?: string;
  end_time?: string;
}

export interface ResourceAssignmentUpdate {
  role?: string;
  start_time?: string;
  end_time?: string;
}

export async function getResourceAssignments(resourceId: number): Promise<ResourceAssignmentOut[]> {
  return apiRequest<ResourceAssignmentOut[]>(`/resources/${resourceId}/assignments`);
}

export async function createResourceAssignment(data: ResourceAssignmentCreate): Promise<ResourceAssignmentOut> {
  return apiRequest<ResourceAssignmentOut>("/resources/assignments", { method: "POST", body: data });
}

export async function updateResourceAssignment(assignmentId: number, data: ResourceAssignmentUpdate): Promise<ResourceAssignmentOut> {
  return apiRequest<ResourceAssignmentOut>(`/resources/assignments/${assignmentId}`, { method: "PATCH", body: data });
}

export async function deleteResourceAssignment(assignmentId: number): Promise<void> {
  return apiRequest<void>(`/resources/assignments/${assignmentId}`, { method: "DELETE" });
}
